csi/src/node/nvmfutil.rs

```diff
@@ -0,0 +1,178 @@
+                now.elapsed()
+            );
+            return Some(path);
+        }
+        retries += 1;
+    }
+    debug!("wait_for_path_to_exist timed out after {:?}", now.elapsed());
+    None
+}
+
+fn nvmeadm_attach_disk(
+    ip_addr: &str,
+    port: u32,
+    nqn: &str,
+    uri: &str,
+) -> Result<String, String> {
+    trace!(
+        "nvmeadm_attach_disk ip_addr={} port={} nqn={}",
+        ip_addr,
+        port,
+        nqn
+    );
+
+    let result = nvmeadm::nvmf_discovery::connect(ip_addr, port, nqn);
+    if let Err(e) = result {
+        let mut in_progress = false;
+        if let Some(ioerr) = e.downcast_ref::<io::Error>() {
+            if let Some(errcode) = ioerr.raw_os_error() {
+                if errcode == 114 {
+                    in_progress = true;
+                    debug!("nvmeadm operation already in progress for {}", uri);
+                }
+            }
+        }
+        if !in_progress {
+            debug!("nvmeadm connect failed error {} {:?}", uri, e);
+            return Err(format!("{}", e));
+        }
+    }
+
+    match wait_for_path_to_exist(uuid_from_str(uri), 10) {
+        Some(path) => Ok(path),
+        _ => {
+            debug!("nvmeadm nvmf device path not found.");
+            Err("No path for nvme device".to_string())
+        }
+    }
+}
+
+pub fn nvmeadm_detach_disk(nqn: &str) -> Result<(), String> {
+    match nvmeadm::nvmf_discovery::disconnect(&nqn) {
+        Ok(_) => {
+            trace!("nvmf disconnected {}", nqn);
+            Ok(())
+        }
+        Err(e) => {
+            debug!("nvmf disconnect {} FAILED.", nqn);
+            Err(format!("{}", e))
+        }
+    }
+}
+
+pub fn nvmf_attach_disk(nvmf_uri: &str) -> Result<String, String> {
+    trace!("nvmf_attach_disk {}", nvmf_uri);
+
+    if let Some(path) = wait_for_path_to_exist(uuid_from_str(nvmf_uri), 1) {
+        return Ok(path);
+    }
+
+    if let Ok(url) = url::Url::parse(nvmf_uri) {
+        if url.scheme() == "nvmf" {
+            let tokens =
+                url.path_segments().map(|c| c.collect::<Vec<_>>()).unwrap();
+            return nvmeadm_attach_disk(
+                url.host_str().unwrap(),
+                u32::from(url.port().unwrap()),
+                tokens[0],
+                nvmf_uri,
+            );
+        }
+    }
+
+    Err(format!("Invalid nvmf URI {}", nvmf_uri))
+}
+
+/// Search for and return path to the device on which a nexus iscsi
+/// target matching the volume id has been mounted or None.
+pub fn nvmf_find(uuid: &str) -> Option<String> {
+    trace!("nvmf_find {}", uuid);
+    match find_nvmf_device_by_uuid(uuid) {
+        Ok(path) => {
+            trace!("nvmf_find for {} got {}", uuid, path);
+            Some(path)
+        }
+        _ => {
+            debug!("nvmf_find for {} FAILED", uuid);
+            None
+        }
+    }
+}
+
+pub fn nvmf_detach_disk(uuid: &str) -> Result<(), String> {
+    // Ugh! hardcoded nqn, bad, bad, bad
```

Comment:
I'd rather see a todo comment here :)

Reply:
Colin will be refactoring the nbd,iscsi and nvmf code to be more object oriented, so I opted to leave the resolution of this to that change.